# Intro to scientific Python programming
**Jan 17, 2015**








### This is a very quick intro to Python programming

 * variables for numbers, lists, and arrays

 * while loops and for loops

 * functions

 * if tests

 * plotting

 * files

Method: show program code through math examples


## Variables, loops, lists, and arrays
<div id="basics:basic:objects"></div>


<p></p>
<img src="fig-bumpy/implementation5.jpg" width=500>



### Do you have access to Python?
<div id="basics:accesspy"></div>

Many methods:

 * Mac and Windows: [Anaconda](https://store.continuum.io/cshop/anaconda/)

 * Ubuntu: `sudo apt-get install`

 * Web browser ([Wakari](https://wakari.io/) or [SageMathCloud](https://cloud.sagemath.com/))

See [How to access Python for doing scientific computing](http://hplgit.github.io/edu/accesspy/accesspy.html) for more details!

### Mathematical example

Most examples will involve this formula:

<!-- Equation labels as ordinary links -->
<div id="basics:seq"></div>

$$
\begin{equation}
\label{basics:seq} \tag{1}
s = v_0t + \frac{1}{2}at^2
\end{equation}
$$

We may view $s$ as a function of $t$: $s(t)$, and also include the
parameters in the notation: $s(t;v_0,a)$.

### A program for evaluating a formula
<div id="basics:formula:eval"></div>


**Task.**

Compute $s$ for $t=0.5$, $v_0=2$, and $a=0.2$.



**Python code.**

In [1]:
t = 0.5
v0 = 2
a = 0.2
s = v0*t + 0.5*a*t**2
print s


1.025


**Execution.**

        Terminal> python distance.py
        1.025


### Assignment statements assign a name to an object

In [5]:
t = 0.5                   # real number makes float object
v0 = 2                    # integer makes int object
a = 0.2                   # float object
s = v0*t + 0.5*a*t**2     # float object
print s

1.025


Rule:

 * evaluate right-hand side; it results in an *object*

 * left-hand side is a name for that object

### Formatted output with text and numbers
<div id="basics:printf"></div>

 * Task: write out text with a number (3 decimals): `s=1.025`

 * Method: printf syntax

In [6]:
print 's=%g' % s       # g: compact notation
print 's=%.2f' % s     # f: decimal notation, .2f: 2 decimals


s=1.025
s=1.02


Modern alternative: format string syntax

In [7]:
print 's={s:.2f}'.format(s=s)


s=1.02


### Programming with a while loop
<div id="basics:while"></div>


 * Task: write out a table of $t$ and $s(t)$ values (two columns),
   for $t\in [0,2]$ in steps of 0.1

 * Method: while loop

In [8]:
v0 = 2
a = 0.2
dt = 0.1  # Increment
t = 0     # Start value
while t <= 2:
    s = v0*t + 0.5*a*t**2
    print t, s
    t = t + dt


0 0.0
0.1 0.201
0.2 0.404
0.3 0.609
0.4 0.816
0.5 1.025
0.6 1.236
0.7 1.449
0.8 1.664
0.9 1.881
1.0 2.1
1.1 2.321
1.2 2.544
1.3 2.769
1.4 2.996
1.5 3.225
1.6 3.456
1.7 3.689
1.8 3.924
1.9 4.161


### Output of the previous program

        Terminal> python while.py
        0 0.0
        0.1 0.201
        0.2 0.404
        0.3 0.609
        0.4 0.816
        0.5 1.025
        0.6 1.236
        0.7 1.449
        0.8 1.664
        0.9 1.881
        1.0 2.1
        1.1 2.321
        1.2 2.544
        1.3 2.769
        1.4 2.996
        1.5 3.225
        1.6 3.456
        1.7 3.689
        1.8 3.924
        1.9 4.161


### Structure of a while loop

In [9]:
while condition:
    <intented statement>
    <intented statement>
    <intented statement>


SyntaxError: invalid syntax (<ipython-input-9-56997f356a2c>, line 2)

Note:

 * the colon in the first line

 * all statements in the loop *must be indented* 
   (no braces as in C, C++, Java, ...)

 * `condition` is a boolean expression (e.g., `t <= 2`)

### Let's take a closer look at the output of our program

        Terminal> python while.py
        0 0.0
        0.1 0.201
        0.2 0.404
        ...
        1.8 3.924
        1.9 4.161


The last line contains 1.9, but the while loop should run also when
$t=2$ since the test is `t <= 2`. Why is this test `False`?

### Let's examine the program in the Python Online Tutor

[Python Online Tutor](http://pythontutor.com/visualize.html#mode=edit):
step through the program and examine variables

In [10]:
a = 0
da = 0.4
while a <= 1.2:
    print a
    a = a + da


0
0.4
0.8


### Ooops, why is `a <= 1.2` when `a` is `1.2`? Round-off errors!

In [11]:
a = 0
da = 0.4
while a <= 1.2:
    print a
    a = a + da
    # Inspect all decimals in da and a
    print 'da=%.16E\na=%.16E' % (da, a)
    print 'a <= 1.2: %g' %  (a <= 1.2)


0
da=4.0000000000000002E-01
a=4.0000000000000002E-01
a <= 1.2: 1
0.4
da=4.0000000000000002E-01
a=8.0000000000000004E-01
a <= 1.2: 1
0.8
da=4.0000000000000002E-01
a=1.2000000000000002E+00
a <= 1.2: 0


**Box.**

Small round-off error in `da` makes `a = 1.2000000000000002`



### Rule: never `a == b` for real `a` and `b`! Always use a tolerance!

In [12]:
a = 1.2
b = 0.4 + 0.4 + 0.4
boolean_condition1 = a == b              # may be False

# This is the way to do it
tol = 1E-14
boolean_condition2 = abs(a - b) < tol    # True


### A list collects several objects in a given sequence
<div id="basics:list"></div>


A list of numbers:

In [13]:
L = [-1, 1, 8.0]


A list can contain any type of objects, e.g.,

In [14]:
L = ['mydata.txt', 3.14, 10]   # string, float, int


Some basic list operations:

In [16]:
L = ['mydata.txt', 3.14, 10]
print L[0]    # print first element (index 0)
print L[1]    # print second element (index 1)
#del L[0]      # delete the first element
print L
print len(L)  # length of L
L.append(-1)  # add -1 at the end of the list
print L

mydata.txt
3.14
['mydata.txt', 3.14, 10]
3
['mydata.txt', 3.14, 10, -1]


### Store our table in two lists, one for each column

In [1]:
v0 = 2
a = 0.2
dt = 0.1  # Increment
t = 0
t_values = []
s_values = []
while t <= 2:
    s = v0*t + 0.5*a*t**2
    t_values.append(t)
    s_values.append(s)
    t = t + dt
print s_values  # Just take a look at a created list

# Print a nicely formatted table
i = 0
while i <= len(t_values)-1:
    print '%.2f  %.4f' % (t_values[i], s_values[i])
    i += 1   # Same as i = i + 1


### For loops
<div id="basics:for"></div>


A for loop is used for visiting elements in a list, one by one:

In [1]:
L = [1, 4, 8, 9]
for e in L:
    print e

Demo in the Python Online Tutor:

In [1]:
list1 = [0, 0.1, 0.2]
list2 = []
for element in list1:
    p = element + 2
    list2.append(p)
print list2


### Traditional for loop: integer counter over list/array indices

In [1]:
somelist = ['file1.dat', 22, -1.5]

for i in range(len(somelist)):
    # access list element through index
    print somelist[i]


Note:

 * `range` returns a list of integers

 * `range(a, b, s)` returns the integers
   `a, a+s, a+2*s, ...` up to *but not including* (!!) `b`

 * `range(b)` implies `a=0` and `s=1`

 * `range(len(somelist))` returns `[0, 1, 2]`

### Let's replace our while loop by a for loop

In [1]:
v0 = 2
a = 0.2
dt = 0.1  # Increment
t_values = []
s_values = []
n = int(round(2/dt)) + 1  # No of t values
for i in range(n):
    t = i*dt
    s = v0*t + 0.5*a*t**2
    t_values.append(t)
    s_values.append(s)
print s_values  # Just take a look at a created list

# Make nicely formatted table
for t, s in zip(t_values, s_values):
    print '%.2f  %.4f' % (t, s)

# Alternative implementation
for i in range(len(t_values)):
    print '%.2f  %.4f' % (t_values[i], s_values[i])


### Traversal of multiple lists at the same time with `zip`

In [1]:
for e1, e2, e3, ... in zip(list1, list2, list3, ...):


Alternative: loop over a common index for the lists

In [1]:
for i in range(len(list1)):
    e1 = list1[i]
    e2 = list2[i]
    e3 = list3[i]
    ...


### Arrays are computationally efficient lists of numbers
<div id="basics:numpy"></div>


 * Lists collect a set of objects in a single variable

 * Lists are very flexible (can grow, can contain "anything")

 * Array: computationally efficient and convenient list

 * Arrays must have fixed length and can only contain numbers of
   the same type (integers, real numbers, complex numbers)

 * Arrays require the `numpy` module

### Examples on using arrays

In [1]:
import numpy
L = [1, 4, 10.0]    # List of numbers
a = numpy.array(L)  # Convert to array
print a
print a[1]          # Access element through indexing
print a[0:2]        # Extract slice (index 2 not included!)
print a.dtype       # Data type of an element
b = 2*a + 1         # Can do arithmetics on arrays
print b

### `numpy` functions creates entire arrays at once

Apply $\ln$ to all elements in array `a`:

In [1]:
c = numpy.log(a)
print c

Create $n+1$ uniformly distributed coordinates in $[a,b]$:

In [1]:
t = numpy.linspace(a, b, n+1)


Create array of length $n$ filled with zeros:

In [1]:
t = numpy.zeros(n)
s = numpy.zeros_like(t)  # zeros with t's size and data type


### Let's use arrays in our previous program

In [1]:
import numpy
v0 = 2
a = 0.2
dt = 0.1  # Increment
n = int(round(2/dt)) + 1  # No of t values

t_values = numpy.linspace(0, 2, n+1)
s_values = v0*t + 0.5*a*t**2

# Make nicely formatted table
for t, s in zip(t_values, s_values):
    print '%.2f  %.4f' % (t, s)


Note: no explicit loop for computing `s_values`!

### Standard mathematical functions are found in the `math` module
<div id="basics:math"></div>

In [1]:
import math
print math.sin(math.pi)

Get rid of the `math` prefix:

In [1]:
from math import sin, pi
print sin(pi)

# Or import everything from math
from math import *
print sin(pi), log(e), tanh(0.5)


### Use the `numpy` module for standard mathematical functions applied to arrays


Matlab users can do

In [1]:
from numpy import *
x = linspace(0, 1, 101)
y = exp(-x)*sin(pi*x)


The Python community likes

In [1]:
import numpy as np
x = np.linspace(0, 1, 101)
y = np.exp(-x)*np.sin(np.pi*x)


Our convention: use `np` prefix, but not in formulas involving
math functions

In [1]:
import numpy as np
x = np.linspace(0, 1, 101)

from numpy import sin, exp, pi
y = exp(-x)*sin(pi*x)


### Plotting
<div id="basics:plot"></div>


Plotting is done with `matplotlib`:

In [18]:
#%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt

v0 = 0.2
a = 2
n = 21  # No of t values for plotting

t = np.linspace(0, 2, n+1)
s = v0*t + 0.5*a*t**2

plt.plot(t, s)
plt.savefig('myplot.png')
plt.show()


ImportError: No module named numpy

### Plotting of multiple curves

In [1]:
import numpy as np
import matplotlib.pyplot as plt

v0 = 0.2
n = 21  # No of t values for plotting

t = np.linspace(0, 2, n+1)
a = 2
s0 = v0*t + 0.5*a*t**2
a = 3
s1 = v0*t + 0.5*a*t**2

plt.plot(t, s0, 'r-',  # Plot s0 curve with red line
         t, s1, 'bo')  # Plot s1 curve with blue circles
plt.xlabel('t')
plt.ylabel('s')
plt.title('Distance plot')
plt.legend(['$s(t; v_0=2, a=0.2)$', '$s(t; v_0=2, a=0.8)$'],
           loc='upper left')
plt.savefig('myplot.png')
plt.show()


## Functions and branching
<div id="basics:func:branching"></div>


<p></p>
<img src="fig-bumpy/branching3.jpg" width=500>



### Functions
<div id="basics:func"></div>

 * $s(t)=v_0t + \frac{1}{2}at^2$ is a mathematical function

 * Can implement $s(t)$ as a Python function `s(t)`

In [1]:
def s(t):
    return v0*t + 0.5*a*t**2

v0 = 0.2
a = 4
value = s(3)   # Call the function


Note:

 * functions start with the keyword `def`

 * statements belonging to the function must be indented

 * function input is represented by arguments
   (separated by comma if more than one)

 * function output is returned to the calling code

 * `v0` and `a` are *global variables*, which
   must be initialized before `s(t)` is called

### Functions can have multiple arguments

`v0` and `a` as function arguments instead of global variables:

In [1]:
def s(t, v0, a):
    return v0*t + 0.5*a*t**2

value = s(3, 0.2, 4)   # Call the function

# More readable call
value = s(t=3, v0=0.2, a=4)


### Keyword arguments are arguments with default values

In [1]:
def s(t, v0=1, a=1):
    return v0*t + 0.5*a*t**2

value = s(3, 0.2, 4)         # specify new v0 and a
value = s(3)                 # rely on v0=1 and a=1
value = s(3, a=2)            # rely on v0=1
value = s(3, v0=2)           # rely on a=1
value = s(t=3, v0=2, a=2)    # specify everything
value = s(a=2, t=3, v0=2)    # any sequence allowed


* Arguments without the argument name are called *positional arguments*

 * Positional arguments must always be listed before the *keyword arguments*
   in the function and in any call

 * The sequence of the keyword arguments can be arbitrary

### Vectorization speeds up the code


Scalar code (work with one number at a time):

In [1]:
def s(t, v0, a):
    return v0*t + 0.5*a*t**2

for i in range(len(t)):
    s_values[i] = s(t_values[i], v0, a)


Vectorized code: apply `s` to the entire array

In [1]:
s_values = s(t_values, v0, a)


How can this work?

 * Expression: v0*t + 0.5*a*t**2 with array `t`

 * `r1 = v0*t`  (scalar times array)

 * `r2 = t**2`  (square each element)

 * `r3 = 0.5*a*r2` (scalar times array)

 * `r1 + r3` (add each element)

### Python functions written for scalars normally work for arrays too!

True if computations involve arithmetic operations and math functions:

In [1]:
from math import exp, sin

def f(x):
    return 2*x + x**2*exp(-x)*sin(x)

v = f(4)  # f(x) works with scalar x

# Redefine exp and sin with their vectorized versions
from numpy import exp, sin, linspace
x = linspace(0, 4, 100001)
v = f(x)  # f(x) works with array x


### Python functions can return multiple values

Return $s(t)=v_0t+\frac{1}{2}at^2$ and $s'(t)=v_0 + at$:

In [1]:
def movement(t, v0, a):
    s = v0*t + 0.5*a*t**2
    v = v0 + a*t
    return s, v

s_value, v_value = movement(t=0.2, v0=2, a=4)


`return s, v` means that we return a *tuple* ($\approx$ list):

In [1]:
def f(x):
    return x+1, x+2, x+3
r = f(3)     # Store all three return values in one object r
print r
type(r)      # What type of object is r?
print r[1]

Tuples are constant lists (cannot be changed)

### A more general mathematical formula (part I)
<div id="basics:formula:piecewise"></div>

Equations from basic kinematics:

$$
\begin{align*}
v = \frac{ds}{dt},\quad s(0)=s_0\\ 
a = \frac{dv}{dt},\quad v(0)=v_0
\end{align*}
$$

Integrate to find $v(t)$:

$$
\int_0^t a(t)dt = \int_0^t \frac{dv}{dt} dt
$$

which gives

$$
v(t) = v_0 + \int_0^t a(t)dt
$$

### A more general mathematical formula (part II)

Integrate again over $[0,t]$ to find $s(t)$:

$$
s(t) = s_0 + v_0t + \int_0^t\left( \int_0^t a(t)dt \right) dt
$$

Example: $a(t)=a_0$ for $t\in[0,t_1]$, then $a(t)=0$ for $t>t_1$:

$$
s(t) = \left\lbrace\begin{array}{ll}
s_0 + v_0 t + \frac{1}{2}a_0 t^2,& t\leq t_1\\ 
s_0 + v_0t_1 + \frac{1}{2}a_0 t_1^2 + a_0t_1(t-t_1),& t> t_1
\end{array}\right.
$$

Need *if test* to implement this!

### Basic if-else tests
<div id="basics:if"></div>


An if test has the structure

In [1]:
if condition:
    <statements when condition is True>
else:
    <statements when condition is False>


Here,

 * `condition` is a boolean expression with value `True` or
   `False`.

In [1]:
if t <= t1:
    s = v0*t + 0.5*a0*t**2
else:
    s = v0*t + 0.5*a0*t1**2 + a0*t1*(t-t1)


### Multi-branch if tests

In [1]:
if condition1:
    <statements when condition1 is True>
elif condition2:
    <statements when condition1 is False and condition2 is True>
elif condition3:
    <statements when condition1 and conditon 2 are False
     and condition3 is True>
else:
    <statements when condition1/2/3 all are False>


Just if, no else:

In [1]:
if condition:
    <statements when condition is True>


### Implementation of a piecewisely defined function with if

A Python function implementing the mathematical function

$$
s(t) = \left\lbrace\begin{array}{ll}
s_0 + v_0 t + \frac{1}{2}a_0 t^2,& t\leq t_1\\ 
s_0 + v_0t_1 + \frac{1}{2}a_0 t_1^2 + a_0t_1(t-t_1),& t> t_1
\end{array}\right.
$$

reads

In [1]:
def s_func(t, v0, a0, t1):
    if t <= t1:
        s = v0*t + 0.5*a0*t**2
    else:
        s = v0*t + 0.5*a0*t1**2 + a0*t1*(t-t1)
    return s


### Python functions containing if will not accept array arguments

In [1]:
def f(x): return x if x < 1 else 2*x
import numpy as np
x = np.linspace(0, 2, 5)
f(x)

Problem: `x < 1` evaluates to a boolean array, not just a boolean

### Remedy 1: Call the function with scalar arguments

In [1]:
n = 201  # No of t values for plotting
t1 = 1.5

t = np.linspace(0, 2, n+1)
s = np.zeros(n+1)
for i in range(len(t)):
    s[i] = s_func(t=t[i], v0=0.2, a0=20, t1=t1)


Can now easily plot:

In [1]:
plt.plot(t, s, 'b-')
plt.plot([t1, t1], [0, s_func(t=t1, v0=0.2, a0=20, t1=t1)], 'r--')
plt.xlabel('t')
plt.ylabel('s')
plt.savefig('myplot.png')
plt.show()


### Remedy 2: Vectorize the if test with `where`


Functions with if tests require a complete rewrite to work with
arrays.

In [1]:
s = np.where(condition, s1, s2)


Explanation:

 * `condition`: array of boolean values

 * `s[i] = s1[i]` if `condition[i]` is `True`

 * `s[i] = s2[i]` if `condition[i]` is `False`

Our example then becomes

In [1]:
s = np.where(t <= t1,
             v0*t + 0.5*a0*t**2,
             v0*t + 0.5*a0*t1**2 + a0*t1*(t-t1))


Note that `t <= t1` with array `t` and scalar `t1` results in a boolean
array `b` where `b[i] = t[i] <= t1`.

### Remedy 3: Vectorize the if test with array indexing

 * Let `b` be a boolean array (e.g., `b = t <= t1`)

 * `s[b]` selects all elements `s[i]` where `b[i]` is `True`

 * Can assign some array expression `expr` of length
   `len(s[b])` to `s[b]`: `s[b] = (expr)[b]`

Our example can utilize this technique with `b` as `t <= t1` and `t > t1`:

In [1]:
s = np.zeros_like(t)  # Make s as zeros, same size & type as t
s[t <= t1] = (v0*t + 0.5*a0*t**2)[t <= t1]
s[t > t1]  = (v0*t + 0.5*a0*t1**2 + a0*t1*(t-t1))[t > t1]


## Files
<div id="basics:files"></div>


<p></p>
<img src="fig-bumpy/files1.jpg" width=400>



### File reading

Put input data in a text file:

        v0 = 2
        a = 0.2
        dt = 0.1
        interval = [0, 2]


**Box.**

How can we read this file into variables `v0`, `a`, `dt`, and `interval`?



### Code for reading files with lines `variable = value`

In [1]:
infile = open('.input.dat', 'r')
for line in infile:
    # Typical line: variable = value
    variable, value = line.split('=')
    variable = variable.strip()  # remove leading/traling blanks
    if variable == 'v0':
        v0 = float(value)
    elif variable == 'a':
        a = float(value)
    elif variable == 'dt':
        dt = float(value)
    elif variable == 'interval':
        interval = eval(value)
infile.close()


### Splitting lines into words is a frequent operation

In [1]:
line = 'v0 = 5.3'
variable, value = line.split('=')
variable
value
variable.strip()  # strip away blanks

Note: must convert `value` to `float` before we can compute with
the value!

### The magic `eval` function

`eval(s)` executes a string `s` as a Python expression and creates the
corresponding Python object

In [1]:
obj1 = eval('1+2')   # Same as obj1 = 1+2
obj1, type(obj1)
obj2 = eval('[-1, 8, 10, 11]')
obj2, type(obj2)
from math import sin, pi
x = 1
obj3 = eval('sin(pi*x)')
obj3, type(obj3)

Why is this so great? We can read formulas, lists, expressions as
text from file and with `eval` turn them into live Python objects!

### Implementing a calculator in Python

Demo:

        Terminal> python calc.py "1 + 0.5*2"
        2.0
        Terminal> python calc.py "sin(pi*2.5) + exp(-4)"
        1.0183156388887342


Just 5 lines of code:

In [1]:
import sys
command_line_expression = sys.argv[1]
from math import *   # Define sin, cos, exp, pi, etc.
result = eval(command_line_expression)
print result


### Modern Python often applies the `with` statement for file handling

In [1]:
with open('.input.dat', 'r') as infile:
    for line in infile:
        ...


No need to close the file when using `with`

### File writing

 * We have $t$ and $s(t)$ values in two lists, `t_values` and
   `s_values`

 * Task: write these lists as a nicely formatted table in a file

Code:

In [1]:
outfile = open('table1.dat', 'w')
outfile.write('# t    s(t)\n')  # write table header
for t, s in zip(t_values, s_values):
    outfile.write('%.2f  %.4f\n' % (t, s))


### Simplified writing of tabular data to file via `numpy.savetxt`

In [1]:
import numpy as np
# Make two-dimensional array of [t, s(t)] values in each row
data = np.array([t_values, s_values]).transpose()

# Write data array to file in table format
np.savetxt('table2.dat', data, fmt=['%.2f', '%.4f'],
           header='t   s(t)', comments='# ')


`table2.dat`:

        # t   s(t)
        0.00 0.0000
        0.10 0.2010
        0.20 0.4040
        0.30 0.6090
        0.40 0.8160
        0.50 1.0250
        0.60 1.2360
        ...
        1.90 4.1610
        2.00 4.4000


### Simplified reading of tabular data from file via `numpy.loadtxt`

In [1]:
data = np.loadtxt('table2.dat', comments='#')


Note:

 * Lines beginning with the comment character `#` are skipped in the reading

 * `data` is a two-dimensional array: `data[i,0]`
   holds the $t$ value and `data[i,1]` the $s(t)$ value in the `i`-th
   row